### PR TITLE
feat: create custom application form

### PR DESCRIPTION
**Pull request by arian81**
## Summary by CodeRabbit

Given the provided information, the following bullet-point list summarizes the changes in the pull request:

- **New Features**
  - Introduced a new application form with custom validation for event participants.
  - Added a search bar to enhance user navigation within the app.
  - Implemented a new `UniversitySelect` component for dynamic university selection in forms.
  - Added a `MultiSelect` component for improved multi-selection capabilities in forms.

- **Enhancements**
  - Updated the visual presentation of the `Background` component.
  - Improved the layout and styling of the `SocialButtons` and `NavBar` components.
  - Enhanced the `Welcome` page with updated navigation links and footer adjustments.

- **Bug Fixes**
  - Corrected class attribute formatting in the `SocialButtons` component.

- **Refactor**
  - Replaced the `Widget` component with a custom form implementation using `react-hook-form` and `zod`.

- **Documentation**
  - No user-facing documentation updates mentioned.

- **Style**
  - Updated styles for the search bar and various form components.

- **Tests**
  - No user-facing test updates mentioned.

- **Chores**
  - Updated GitHub Actions workflow to use Node.js version 20 and switched to pnpm.
  - Added "src/pages/test.tsx" to `.gitignore` to exclude it from version control.

- **Revert**
  - No reverts mentioned.

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -101,3 +104,50 @@ enum Status {
...
   CHECKED_IN
 }
+
+model DH10Application {
+  id String @id @default(cuid())
+
+  firstName String
+  lastName  String
+
+  birthday DateTime
```
Comment (err53) [inline prisma/schema.prisma:114]: Should this be DateTime? @db.Date would be better I think...

https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#cockroachdb-6

File: src/data/applicationSelectData.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+const degrees: SelectChoice[] = [
+  "Bachelors",
+  "Masters",
+  "PhD",
+  "Associate's",
+  "Certificate",
+  "Diploma",
+].map((v) => ({ value: v, label: v }));
+
+const heardFrom: SelectChoice[] = [
```
Comment (err53) [inline src/data/applicationSelectData.ts:260]: would `referredBy` be a more appropriate name here

**Review by Krish120003 — APPROVED** [review]

File: src/schemas/application.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
+import z from "zod";
+
+const schema = z.object({
+  firstName: z.string().min(1).max(255),
+  lastName: z.string().min(1).max(255),
+  birthday: z.string().refine(
+    (value) => {
+      // parse date
+      const date = new Date(value);
```
Comment (err53) [inline src/schemas/application.ts:9]: does this take into account timezones? i assume serverside date is stored in UTC?

File: src/data/applicationSelectData.ts (hunk 2 above)
Comment (err53) [reply src/data/applicationSelectData.ts:260]: also why is there both heardFrom and discoveredFrom (misspelled as discoverdFrom)

File: src/server/router/application.ts [hunk 4]
```diff
@@ -225,4 +231,196 @@ export const applicationRouter = router({
...
+    }
+    // console.log(autofill)
+
+    return autofill;
+  }),
+  submitDh10: protectedProcedure
+    .input(applicationSchema)
+    .mutation(async ({ ctx, input }) => {
+      // make sure there is no existing application
+      const existing = await ctx.prisma.dH10Application.findFirst({
```
Comment (err53) [inline src/server/router/application.ts:391]: this can lead to a race condition. just use `create` and try/catch, no?

Comment (err53) [reply src/server/router/application.ts:391]: or use the transaction syntax coderabbit suggested

**Review by err53 — CHANGES_REQUESTED** [review]

File: src/server/router/application.ts (hunk 4 above)
Comment (Krish120003) [reply src/server/router/application.ts]: should be fixed

File: src/data/applicationSelectData.ts (hunk 2 above)
Comment (Krish120003) [reply src/data/applicationSelectData.ts:260]: Could you explain this again

Comment (err53) [reply src/data/applicationSelectData.ts:260]: its called heardFrom here but there's also https://github.com/deltahacks/portal/pull/149/files#r1396215969

Comment (arian81) [reply src/data/applicationSelectData.ts]: ok i changed the naming

**Pull request merged**